keps/0029-operator-dependencies.md

```diff
@@ -0,0 +1,282 @@
+
+This KEP is aiming at streamlining this experience for users and developers.
+
+### Goals
+
+Dependencies can be a complex topic. This KEP is not trying to boil the dependency ocean but rather limits itself to installation dependencies only, i.e. a set of `Operators/Instances` being installed together and removed together as a unit.
+
+### Non-Goals
+
+Dependency on an already running `Instance` is a non-goal. It is easy to imagine a situation when a new operator (e.g Kafka) may want to depend on the existing Zookeeper instance. However, such life-cycle dependency presents major challenges e.g. what happens when Zookeeper is removed? What happens when Zookeeper is upgraded, and the new version is incompatible with the current Kafka `Instance`? How can we ensure the compatibility? This KEP deliberately ignores this area and instead focuses on installation dependencies. Additionally, this KEP does not address output variables or referencing `Instance` resources.
+
+## Proposal
+
+KUDO operators **already have** a mechanism to deal with installation dependencies called [plans, phases, and steps](https://kudo.dev/docs/developing-operators/plans.html#overview) with serial or parallel execution strategy. This mechanism is already powerful enough to express any dependency hierarchy including transitive dependencies (see more about this in [implementation details](#implementation-details)). The core of this proposal is to reuse this mechanism and extend it with the ability to install operators.
+
+### Implementation details
+
+#### Operator Task
+
+A new task kind `Operator` is introduced which extends `operator.yaml` with the ability to install dependencies. Let's take a look at the Kafka+Zookeeper example:
+
+```yaml
+apiVersion: kudo.dev/v1beta1
+name: "kafka"
+operatorVersion: "1.3.1"
+kudoVersion: 0.12.0
+kubernetesVersion: 1.14.8
+appVersion: 2.5.0
+url: https://kafka.apache.org/
+tasks:
+  - name: zookeeper-operator
+    kind: Operator
+    spec:
+        package: zookeeper
+
+  ...
+
+plans:
+  deploy:
+    strategy: serial
+    phases:
+      - name: deploy-zookeeper
+        strategy: serial
+        steps:
+          - name: zookeeper
+            tasks:
+              - zookeeper-operator
+      - name: deploy-kafka
+        strategy: serial
+        steps:
+          ...
+```
+
+The `zookeeper-operator` task specification is equivalent to `kudo install zookeeper` CLI command which installs the Zookeeper package from the official repo. Here is a complete `Operator` task specification:
+
+```yaml
+tasks:
+- name: demo
+  kind: Operator
+  spec:
+    package: # required, either repo package name, local package folder or an URL to package tarball
+    repo: # optional, name of local repository configuration to use
```

Comment:
I'm not sure if it's a good idea to include the repo URL in the operator definition. This would open up a lot of ways for people to sneak in unintended operators and have security implications.

If (as an operator user) want to use only operators from `https://official-operators.kudo.dev`, I don't expect a nested operator from `https://unknown-third-party-operator.com/` to be installed.

Additionally, I would expect problems with air-gapped deployments if we allow this.

This should at least be disabled by default and only enabled via an explicit command line flag if we allow this.

It would be nice to have for local development though.

Reply:
> If I (as an operator user) want to use only operators from https://official-operators.kudo.dev, I don't expect a nested operator from https://unknown-third-party-operator.com/ to be installed.

I think it's up to maintainers of `official-operators` to check/enforce that they do not depend on unofficial ones?

> repo names are client controlled and not really standardized

good point, a local repo name does not make any sense in this context

How about just removing this field for now, requiring that the dependent operator+version be available "somehow" to the CLI?

Reply:
> How about just removing this field for now, requiring that the dependent operator+version be available "somehow" to the CLI?

Agree. Initially, I just copied that from the `kudo install` settings but now, I can see, how repo names can be arbitrary and non-standardized between environments. A repo-url might make more sense but let's tackle this later.